File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        curr = root
        res = []
        while curr or stack:
            if curr:
                stack.append(curr)
                curr = curr.left
            else:
                top = stack.pop()
                res.append(top.val)
                curr = top.right
        return res
        

File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-5.java
 /**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        var curr = root;

        while (curr != null) {
            if (curr.right == null) {
                result.add(curr.val);
                curr = curr.left;
            } else {
                var prev = curr.right;
                while (prev.left != null && prev.left != curr) {
                    prev = prev.left;
                }

                if (prev.left == null) {
                    prev.left = curr;
                    result.add(curr.val);
                    curr = curr.right;
                } else {
                    prev.left = null;
                    curr = curr.left;
                }
            }
        }

        Collections.reverse(result);
        return result;
    }
}